### PR TITLE
E2615 -Testing User model and Users controller 

### PR DESCRIPTION
**Pull request by AtharvaW195**
## Summary by CodeRabbit

* **Bug Fixes**
  * POST/role endpoints now handle missing passwords and missing roles correctly (returns proper errors instead of failing).

* **Improvements**
  * More flexible password validation and reset behavior; clearer role-based scopes and authorization checks; improved TA→instructor and TA course associations; JWT generation now explicit about algorithm.

* **Tests**
  * Added comprehensive model and API request specs covering users, roles, validations, auth, and edge cases.

* **Chores**
  * Enforced unique team-participant constraint at the database level.

### CODE REVIEW

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -70,6 +68,10 @@ def managed_users
   def role_users
     name = params[:name].split('_').map(&:capitalize).join(' ')
     role = Role.find_by(name:)
+    unless role
```
Comment (aanand-1706) [inline app/controllers/users_controller.rb:71]: `  role = Role.find_by!(name:)`

Isn't this better?

File: app/models/ta.rb [hunk 2]
```diff
@@ -1,23 +1,19 @@
 # frozen_string_literal: true
 
 class Ta < User
+  validates :parent_id, presence: true
```
Comment (aanand-1706) [inline app/models/ta.rb:4]: Could this break existing entries?

File: db/migrate/20260428000000_update_teams_participants_unique_index.rb [hunk 3]
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+class UpdateTeamsParticipantsUniqueIndex < ActiveRecord::Migration[8.0]
+  def change
+    remove_index :teams_participants, name: :index_teams_participants_on_participant_id_unique
```
Comment (aanand-1706) [inline db/migrate/20260428000000_update_teams_participants_unique_index.rb:5]: How was `index_teams_participants_on_participant_id_unique` being used earlier?

File: app/models/ta.rb [hunk 4]
```diff
@@ -1,23 +1,19 @@
...
-  def courses_assisted_with
-    courses = TaMapping.where(ta_id: id)
-    courses.map { |c| Course.find(c.course_id) }
+    parent_id
   end
 
   def courses_assisted_with
-    courses = TaMapping.where(ta_id: id)
-    courses.map { |c| Course.find(c.course_id) }
+    Course.joins(:ta_mappings).where(ta_mappings: { user_id: id }).distinct.to_a
```
Comment (aanand-1706) [inline app/models/ta.rb:17]: Is `user_id` present in `ta_mappings`?
